src/lib/httpserver.rs

```diff
@@ -0,0 +1,143 @@
+    let req_time = Instant::now();
+    let event_type = match req.headers().get("X-GitHub-Event") {
+        Some(hv) => match hv.to_str() {
+            Ok(s) => s.to_owned(),
+            Err(_) => {
+                *res.status_mut() = StatusCode::BAD_REQUEST;
+                return Ok(res);
+            }
+        },
+        None => {
+            *res.status_mut() = StatusCode::BAD_REQUEST;
+            return Ok(res);
+        }
+    };
+
+    let pl = match authenticate(req, &snare).await {
+        Ok(p) => p,
+        Err(_) => {
+            *res.status_mut() = StatusCode::UNAUTHORIZED;
+            return Ok(res);
+        }
+    };
+
+    let (json_str, owner, repo) = match parse(pl) {
+        Ok((j, o, r)) => (j, o, r),
+        Err(_) => {
+            *res.status_mut() = StatusCode::BAD_REQUEST;
+            return Ok(res);
+        }
+    };
+
+    // We now want to find the per-repo program to run while making sure that we aren't tricked into
+    // executing a file outside of the repos dir.
+    let mut p = PathBuf::new();
+    p.push(&snare.config.reposdir);
+    p.push(owner);
+    p.push(repo);
+    if let Ok(p) = p.canonicalize() {
+        if let Some(s) = p.to_str() {
+            if s.starts_with(&snare.config.reposdir) {
+                // We can tolerate the `unwrap` call below because if it fails it means that
+                // something has gone so seriously wrong in the other thread that there's no
+                // likelihood that we can recover.
+                let qj = QueueJob::new(s.to_owned(), req_time, event_type, json_str);
+                (*snare.queue.lock().unwrap()).push_back(qj);
+                *res.status_mut() = StatusCode::OK;
+                // If the write fails, it almost certainly means that the pipe is full i.e. the
+                // runner thread will be notified anyway. Just in case something else very odd
+                // happened, the runner thread periodically checks the queue "just in case", so
+                // even if this write does fail, the job will be picked up in the not too distant
+                // future. In either case, a failed write doesn't stop progress being made.
+                nix::unistd::write(snare.event_write_fd, &[0]).ok();
+                return Ok(res);
+            }
+        }
+    }
+    // We couldn't find a per-repo program for this request.
+    *res.status_mut() = StatusCode::BAD_REQUEST;
+    Ok(res)
+}
+
+/// Authenticate this request. If successful, return the body of the request. Any error at all
+/// leads to a blanket `Err(())` so that an attacker cannot deduce anything about the cause of the
+/// failure to authenticate.
+async fn authenticate(req: Request<Body>, snare: &Arc<Snare>) -> Result<Bytes, ()> {
+    // Extract the string 'def' from "X-Hub-Signature: abc=def"
+    let sig = req
+        .headers()
+        .get("X-Hub-Signature")
+        .ok_or(())?
+        .to_str()
+        .map_err(|_| ())?
```

Comment:
Since performance isn't key for snare, why don't we make the error type `Box<Error>`, then you can kill all of the `map_err`s.

Reply:
It's partly to emphasise that "this is not an error you can recover from because you can't even tell what it was".

Reply:
OK, fair enough.